### PR TITLE
Fix #5312: Make todo open check locally runnable

### LINKED ISSUE #5312: [Feature Request]: TODO format checker should be run in static_checks.sh & be runnable locally
### Is your feature request related to a problem? Please describe.

The TODO script currently requires using the gh tool locally to download the list of open issues before comparing them in the codebase:

https://github.com/oppia/oppia-android/blob/bb69b2264903d75b305e96864c3c37d3fba261e2/scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt#L27-L29.

### Describe the solution you'd like

The script is possible to be made runnable locally without needing to interact with gh ahead of time. This is ideal as it's both easier to do, and makes it simpler to integrate in static_checks.sh (so that developers can check it ahead of time without needing to rely on CI).

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation
Fixes #5312

This updates the TODO open check script to be locally runnable rather than requiring the developer to manually download the list of issues from GitHub to analyze. This simplifies things and allows the script to be easily run within the ``static_checks.sh`` script. This is being done via interacting directly with GitHub using its RESTful API (see https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#list-repository-issues) in conjunction with the user's local auth token used to set up their ``gh`` tool (which needs to be set up, hence the changes to the wiki documentation and clear error messages from the new ``GitHubClient`` utility).

To further simplify things, a ``regenerate`` mode was added to regenerate the TODO exemptions textproto file (which is helpful for #4929 hence why this comes before that PR).

The new command syntax to perform the TODO check is:

```sh
bazel run //scripts:todo_open_check -- <path_to_dir_root> <path_to_proto_binary> [regenerate]
```

With a specific example to just perform the check:

```sh
bazel run //scripts:todo_open_check -- $(pwd) scripts/assets/todo_open_exemptions.pb
```

And an example to also perform regeneration:

```sh
bazel run //scripts:todo_open_check -- $(pwd) scripts/assets/todo_open_exemptions.pb regenerate
```

Some other things specifically to note:
- TODO exemptions needed to be updated in this PR due to TODO utility & test changes. The new file was created using the new regenerate functionality.
- The TODO check has been added to the end of ``static_checks.sh``.
- The GitHub CI workflow was updated to use the new script syntax appropriately.
- This is the first time scripts have been updated to integrate with Retrofit, and this setup is going to be reused in the future for other services.
- The data model for issues has been updated to better represent the remote data structure.
- Moshi is being used along with Retrofit for an easier interaction with GitHub as a remote endpoint. All of this has been wrapped in a new ``GitHubClient``.
- ``GitHubClient`` is designed to download all issues regardless of length (whereas before the manual download step was limited to the first 2000 issues of the repository) using pagination.
- New tests were added to verify the regenerate flow (and properly set up the mock OkHttp server since the script now relies on an HTTP endpoint to download the GitHub issues itself).
- ``GitHubIssue`` is exempted from tests since it's just a basic data structure, so there's no specific logic to test.
- ``GitHubService`` is exempted from tests since it's a template to generate code via Retrofit's annotation processor, so there's no specific logic to test.
- All scripts proto libraries were updated to use normal Java (rather than Java lite) generation to provide text format support.
- The file paths included in ``TodoOpenCheck``'s output has been simplified to be relative to the repository root rather than using absolute paths (for parity with many other app scripts).
- Since ``GitHubClient``'s tests required interacting with ``gh``, a new ``FakeCommandExecutor`` was added (along with its own tests) to provide support for orchestrating local utilities. This may be useful in other tests in the future, though some of those script tests intentionally integrate with environment commands like ``git`` and ``bazel``.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
This is an infrastructure-only PR.

### CODE REVIEW

File: scripts/assets/test_file_exemptions.textproto
Comment (BenHenning) [inline scripts/assets/test_file_exemptions.textproto:751]: TODO: Document these in description.

File: scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt:30]: TODO: Document.

File: scripts/src/java/org/oppia/android/scripts/common/remote/GitHubService.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/remote/GitHubService.kt:22]: Include in PR description.

File: scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel:18]: TODO: Document these changes.

File: wiki/Static-Analysis-Checks.md
Comment (BenHenning) [inline wiki/Static-Analysis-Checks.md]: TODO: Finish documentation.

File: wiki/Static-Analysis-Checks.md
Comment (BenHenning) [inline wiki/Static-Analysis-Checks.md:177]: TODO: Cover this in PR description.

File: scripts/static_checks.sh
Comment (BenHenning) [inline scripts/static_checks.sh:114]: TODO: Mention changes in description.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt:25]: TODO: Document new script changes.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt]: TODO: Add tests for this bit.

File: scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt:107]: TODO: Document this output format difference.

File: scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt:87]: TODO: Propagate ``e``.

File: scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt:52]: TODO: Add tests.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed changes.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt]: Added tests in latest changes.

File: scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt:52]: Added tests in latest changes.

File: scripts/src/java/org/oppia/android/scripts/common/testing/FakeCommandExecutor.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/testing/FakeCommandExecutor.kt:21]: Add PR docs for this.

File: wiki/Static-Analysis-Checks.md
Comment (BenHenning) [inline wiki/Static-Analysis-Checks.md]: Finish this.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes.

File: wiki/Static-Analysis-Checks.md
Comment (BenHenning) [reply wiki/Static-Analysis-Checks.md]: Fixed in latest changes.

File: scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt:87]: Fixed in latest changes.

**Review by BenHenning — COMMENTED** [review]
Reviewed latest changes.

File: scripts/assets/test_file_exemptions.textproto
Comment (BenHenning) [reply scripts/assets/test_file_exemptions.textproto:751]: Added to PR description.

File: scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/GitHubClient.kt:30]: Docs added & mentioned in PR description.

File: scripts/src/java/org/oppia/android/scripts/common/remote/GitHubService.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/remote/GitHubService.kt:22]: Included.

File: scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/proto/BUILD.bazel:18]: Included in PR description.

File: wiki/Static-Analysis-Checks.md
Comment (BenHenning) [reply wiki/Static-Analysis-Checks.md]: Finished documentation in latest changes.

File: wiki/Static-Analysis-Checks.md
Comment (BenHenning) [reply wiki/Static-Analysis-Checks.md:177]: Updated PR description.

File: scripts/static_checks.sh
Comment (BenHenning) [reply scripts/static_checks.sh:114]: Included in description.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/todo/TodoOpenCheck.kt:25]: Included in description.

File: scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt:107]: Updated PR description.

File: scripts/src/java/org/oppia/android/scripts/common/testing/FakeCommandExecutor.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/testing/FakeCommandExecutor.kt:21]: Updated PR description.

**Comment by BenHenning** [comment]
PR looks good to me from my own reviews; just waiting on CI before assigning this for review.

**Comment by BenHenning** [comment]
Looks like there were some legitimate CI failures, so re-running to ensure everything's now stable.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest fix changes; everything seems good.

**Comment by BenHenning** [comment]
CI is passing minus the one expected failed check. This should be good to go into review.

**Comment by BenHenning** [comment]
@adhiamboperes PTAL for codeowners.

File: scripts/src/java/org/oppia/android/scripts/common/model/BUILD.bazel
Comment (adhiamboperes) [inline scripts/src/java/org/oppia/android/scripts/common/model/BUILD.bazel:9]: Isn't this used in scripts "production" code as well? Same to `github_service` below.

File: scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt
Comment (adhiamboperes) [inline scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt]: We could make this specific to "which" command, since we have a [script](https://github.com/oppia/oppia-android/wiki/Static-Analysis-Checks#how-to-run-static-checks-locally) for running all static checks locally, and the message could become ambiguous in this scenario.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @BenHenning! I have requested some changes.

File: scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/todo/TodoOpenCheckTest.kt]: I'm not opposed to being clearer in this output--done.

File: scripts/src/java/org/oppia/android/scripts/common/model/BUILD.bazel
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/model/BUILD.bazel:9]: This actually has nothing to do with production vs. tests. It was a hack I introduced back when the first scripts were added to ensure that production _app_ code can't depend on scripts dependencies (since they run in a JVM environment and are allowed completely different transitive dependencies from Android). It's confusing, and semantically wrong.

It will also be fixed in #4920 (which is a ways away from being updated & ready for review).

**Review by BenHenning — COMMENTED** [review]
Updated to latest & addressed comments. PTAL @adhiamboperes.

**Review by adhiamboperes — APPROVED** [review]
Thanks @BenHenning! LGTM

**Comment by BenHenning** [comment]
Since this has been fully approved, #5313 is merged, and this PR's branch is up-to-date with develop, enabling auto-merge.

**Pull request merged**
